### PR TITLE
stagehand.context pages as stagehand pages and supporting more than one page

### PR DESCRIPTION
**Pull request by miguelg719**
# why
Currently, Stagehand's context only supports managing a single page through `stagehand.page`. This limits users who need to work with multiple browser pages simultaneously, which is a common use case in web automation (e.g., comparing data across tabs, handling popups, or parallel processing).

# what changed
1. Enhanced Context Management
Created EnhancedContext interface that properly extends Playwright's BrowserContext
Modified context.newPage() to return our enhanced Page type with act/extract/observe methods
Modified context.pages() to return an array of enhanced Page instances

2. Page Tracking
Added WeakMap in StagehandContext to track StagehandPage mappings
Maintains memory efficiency by using WeakMap for automatic cleanup

3. Type System Improvements
Updated type definitions to properly reflect enhanced page capabilities
Fixed circular dependencies between StagehandPage and StagehandContext
Ensured type safety when working with multiple pages

Now you can call act/extract/observe on new pages:
```typescript
  const mainPage = stagehand.page;
  await mainPage.goto("https://example.com");

  const context = await stagehand.context;
  const newPage = await context.newPage();
  await newPage.goto("https://google.com");

  const observeRes = await newPage.observe({
    instruction: "click the google search button",
  });
```

# test plan
adding deterministic tests

### CODE REVIEW

File: lib/StagehandContext.ts [hunk 1]
```diff
@@ -1,29 +1,112 @@
...
 import { Stagehand } from "./index";
+import { StagehandPage } from "./StagehandPage";
+import { Page } from "../types/page";
+
+// Define the enhanced context type that includes our modified methods
+export interface EnhancedContext
+  extends Omit<PlaywrightContext, "newPage" | "pages"> {
+  newPage(): Promise<Page>;
+  pages(): Page[];
+}
```
Comment (sameelarif) [inline lib/StagehandContext.ts]: should probably move this to the `types` folder and export it from `index.ts`

Comment (miguelg719) [reply lib/StagehandContext.ts]: yup on it

File: lib/StagehandContext.ts [hunk 2]
```diff
@@ -1,29 +1,125 @@
...
+  }
+
+  private async createStagehandPage(
+    page: PlaywrightPage,
+  ): Promise<StagehandPage> {
+    const stagehandPage = await new StagehandPage(
+      page,
+      this.stagehand,
+      this,
+      this.stagehand.getLLMClient(),
```
Comment (kamath) [inline lib/StagehandContext.ts]: would rather just make `llmClient`, `systemPrompt`, etc. `public readonly` access modifier and just do `this.stagehand.llmClient` in `StagehandPage` for consistency

File: evals/deterministic/tests/BrowserContext/multiPage.test.ts [hunk 3]
```diff
@@ -0,0 +1,268 @@
+import { test, expect } from "@playwright/test";
+import { Stagehand } from "@/dist";
+import StagehandConfig from "@/evals/deterministic/stagehand.config";
+import { Page } from "../../../../types/page";
```
Comment (kamath) [inline evals/deterministic/tests/BrowserContext/multiPage.test.ts]: this should import from `dist` to ensure we're importing the built types to avoid a situation in which users can't import the right types

File: lib/StagehandPage.ts [hunk 4]
```diff
@@ -69,7 +83,7 @@ export class StagehandPage {
       },
       on: () => {
         throw new Error(
-          "You seem to be referencing a page in an uninitialized `Stagehand` object. Ensure you are running `await stagehand.init()` on the Stagehand object before referencing the `page` object.",
```
Comment (kamath) [inline lib/StagehandPage.ts:72]: ?

File: lib/StagehandPage.ts [hunk 5]
```diff
@@ -209,13 +223,52 @@ export class StagehandPage {
...
     const page = this.intPage;
     const stagehand = this.stagehand;
-    this.intPage = new Proxy(page, {
-      get: (target, prop) => {
+
+    // Clean up the throwing methods before creating the new proxy
+    delete this.intPage.on;
+    delete this.intPage.act;
+    delete this.intPage.extract;
+    delete this.intPage.observe;
```
Comment (kamath) [inline lib/StagehandPage.ts]: this doesn't look right, feels unsafe

File: lib/index.ts [hunk 6]
```diff
@@ -377,6 +377,35 @@ export class Stagehand {
...
+    return this.userProvidedInstructions;
+  }
+
+  public getAPIClient(): StagehandAPI | undefined {
+    return this.apiClient;
+  }
+
+  public getWaitForCaptchaSolves(): boolean {
+    return this.waitForCaptchaSolves;
+  }
```
Comment (kamath) [inline lib/index.ts]: make these underlying variables `public readonly`

File: lib/index.ts [hunk 7]
```diff
@@ -377,6 +377,35 @@ export class Stagehand {
...
+
+  public setActivePage(page: StagehandPage): void {
+    this.stagehandPage = page;
+  }
```
Comment (kamath) [inline lib/index.ts]: what if we make the input parameter an index? i.e. specify the page in `context.pages()` -- helps prevent edge cases

**Review by kamath — CHANGES_REQUESTED** [review]

File: evals/deterministic/tests/BrowserContext/multiPage.test.ts [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    expect(await stagehand.page.title()).toBe("Page 1");
+
+    // Create another page and verify it becomes active
+    const anotherPage = await stagehand.context.newPage();
+    expect(stagehand.page).toBe(anotherPage);
+    expect(stagehand.page).not.toBe(newPage);
+  });
+
+  /**
+   * Test simultaneous operations on multiple pages THIS WILL FAIL
```
Comment (seanmcguire12) [inline evals/deterministic/tests/BrowserContext/multiPage.test.ts]: delete

**Review by kamath — APPROVED** [review]
